aws_lambda_powertools/event_handler/api_gateway.py

```diff
@@ -1625,9 +1639,19 @@ def swagger_handler():
                 swagger_js = f"{base_path}/swagger.js"
                 swagger_css = f"{base_path}/swagger.css"
 
-            openapi_servers = servers or [Server(url=(base_path or "/"))]
+            spec = get_swagger_spec()
+            body = generate_swagger_html(spec, swagger_js, swagger_css)
```

Comment:
I think we need to add new parameters to this function, right? We serve the Swagger specification in the function [generate_swagger_html](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/event_handler/openapi/swagger_ui/html.py#L1) and currently we are using the field `spec` in the [SwaggerUIOptions](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/event_handler/openapi/swagger_ui/html.py#L36) definition.

Reply:
I added the `get_swagger_spec` function as a way to remove duplicate code between the new route I added and the existing route that serves the documentation UI. No arguments need to be passed because the function is defined with in the `enable_swagger` function (just like the endpoint functions). If I hadn't moved the shared logic to a shared function (`get_swagger_spec`), the code would have looked like:

```python
def enable_swagger(
        self,
        *,
        path: str = "/swagger",
        title: str = "Powertools for AWS Lambda (Python) API",
        version: str = DEFAULT_API_VERSION,
        openapi_version: str = DEFAULT_OPENAPI_VERSION,
        summary: Optional[str] = None,
        description: Optional[str] = None,
        tags: Optional[List[Union["Tag", str]]] = None,
        servers: Optional[List["Server"]] = None,
        terms_of_service: Optional[str] = None,
        contact: Optional["Contact"] = None,
        license_info: Optional["License"] = None,
        swagger_base_url: Optional[str] = None,
        middlewares: Optional[List[Callable[..., Response]]] = None,
        enable_download_spec: bool = False,
    ):
        """
        ...
        """
        ...

        if enable_download_spec:

            @self.get(f"{path}.json", include_in_schema=False)
            def swagger_json():
                # The following 14 lines are duplicated below
                base_path = self._get_base_path()
                openapi_servers = servers or [Server(url=(base_path or "/"))]
                spec = self.get_openapi_json_schema(
                    title=title,
                    version=version,
                    openapi_version=openapi_version,
                    summary=summary,
                    description=description,
                    tags=tags,
                    servers=openapi_servers,
                    terms_of_service=terms_of_service,
                    contact=contact,
                    license_info=license_info,
                )
                return Response(
                    status_code=200,
                    content_type="application/json",
                    body=spec,
                )

        @self.get(path, middlewares=middlewares, include_in_schema=False)
        def swagger_handler():
            base_path = self._get_base_path()

            if swagger_base_url:
                swagger_js = f"{swagger_base_url}/swagger-ui-bundle.min.js"
                swagger_css = f"{swagger_base_url}/swagger-ui.min.css"
            else:
                swagger_js = f"{base_path}/swagger.js"
                swagger_css = f"{base_path}/swagger.css"

            base_path = self._get_base_path()
            openapi_servers = servers or [Server(url=(base_path or "/"))]
            spec = self.get_openapi_json_schema(
                title=title,
                version=version,
                openapi_version=openapi_version,
                summary=summary,
                description=description,
                tags=tags,
                servers=openapi_servers,
                terms_of_service=terms_of_service,
                contact=contact,
                license_info=license_info,
            )
            body = generate_swagger_html(spec, swagger_js, swagger_css)

            return Response(
                status_code=200,
                content_type="text/html",
                body=body,
            )
```

Reply:
> I added the `get_swagger_spec` function as a way to remove duplicate code between the new route I added and the existing route that serves the documentation UI. No arguments need to be passed because the function is defined with in the `enable_swagger` function (just like the endpoint functions). If I hadn't moved the shared logic to a shared function (`get_swagger_spec`), the code would have looked like:

I understand your point and with this implementation, the customer can open a tab/curl/wget and now download the OpenAPI specification requesting `https://www.example.com/swagger.json` directly, this works perfectly.

But what I want to say is: we are implementing the OpenAPI + SwaggerUI specification, where customers can access the swagger definition in the browser (using SwaggerUI) and test endpoints, see models definition and others. Another super common thing in SwaggerUI is to display the link to directly access the JSON/YAML of that OpenAPI spec. If we don't make the modification I suggested, the customer will be able to download OpenAPI by directly accessing `https://www.example.com/swagger.json` (if they know), but this link will not appear in the SwaggerUI interface, which may confuse some customers.
![image](https://github.com/aws-powertools/powertools-lambda-python/assets/4295173/9e530177-bc66-4933-84d3-f2c24293aa6b)

We would do something like this:

**api_gateway.py**
```python
spec = get_swagger_spec()
body = generate_swagger_html(spec, swagger_js, swagger_css, enable_download_spec, path)
```

**html.py**

```python
if enable_download_spec:
    js_json_spec = f'url: "{path}.json",'
else:
    js_json_spec = f"""
    spec: JSON.parse(`
            {spec}
        `.trim()),
    """
```

```js
<script>
  var swaggerUIOptions = {{
    dom_id: "#swagger-ui",
    docExpansion: "list",
    deepLinking: true,
    filter: true,
    layout: "BaseLayout",
    showExtensions: true,
    showCommonExtensions: true,
    {js_json_spec}
    presets: [
      SwaggerUIBundle.presets.apis,
      SwaggerUIBundle.SwaggerUIStandalonePreset
    ],
    plugins: [
      SwaggerUIBundle.plugins.DownloadUrl
    ]
  }}

  var ui = SwaggerUIBundle(swaggerUIOptions)
</script>
```

This code is just an example, we can improve it. Does this make sense to you? What do you think?

Reply:
Ah, I see what you were originally getting at. Yes, I will make an adjustment similar to the one you mentioned. 

From a user experience perspective, it feels a little burdonsome to have to set the `enable_download_spec` parameter to `True` _and_ to add an additional API endpoint if I'm not using `proxy+`. It's doable, but it would be better if the Swagger UI had a link to the spec JSON that just worked for everyone.

Idea for you: what if the `/swagger` endpoint took an optional query string of `?view=json`:

```python
@self.get(path, middlewares=middlewares, include_in_schema=False)
def swagger_handler(view: Annotated[Optional[str], Query()] = None):
```

This query string would be used to conditionally return the documentation as JSON instead of the HTML.

With something like this, every user would get the ability to download the spec, and the download link would be like `https://www.example.com/swagger?view=json`. And in this case, `enable_download_spec` could be removed.

Let me know your thoughts and I will implement the download link one way or the other.